llvm/include/llvm/ADT/Hashing.h

```diff
@@ -322,24 +306,20 @@ struct hash_state {
   }
 };
 
-
-/// A global, fixed seed-override variable.
-///
-/// This variable can be set using the \see llvm::set_fixed_execution_seed
-/// function. See that function for details. Do not, under any circumstances,
-/// set or read this variable.
-extern uint64_t fixed_seed_override;
-
+/// In LLVM_ENABLE_ABI_BREAKING_CHECKS builds, the seed is non-deterministic
+/// (address of a variable) to prevent having users depend on the particular
+/// hash values. On platforms without ASLR, this is still likely
+/// non-deterministic per build.
 inline uint64_t get_execution_seed() {
-  // FIXME: This needs to be a per-execution seed. This is just a placeholder
-  // implementation. Switching to a per-execution seed is likely to flush out
-  // instability bugs and so will happen as its own commit.
-  //
-  // However, if there is a fixed seed override set the first time this is
-  // called, return that instead of the per-execution seed.
-  const uint64_t seed_prime = 0xff51afd7ed558ccdULL;
-  static uint64_t seed = fixed_seed_override ? fixed_seed_override : seed_prime;
-  return seed;
+  // Work around x86-64 negative offset folding for old Clang -fno-pic
+  // https://reviews.llvm.org/D93931
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS &&                                         \
+    (!defined(__clang__) || __clang_major__ > 11)
```

Comment:
Is it an ABI problem that this ifdef exists?  I mean, LLVM libraries built with clang<11 can't be used by programs built with clang>11.  With LLVM_ENABLE_ABI_BREAKING_CHECKS, I guess it's unlikely to cause issues, though.  (I guess you could use an empty inline asm as a workaround if you wanted to.)

Reply:
The clang condition introduced a slight ABI problem when mixing `clang<11` and `clang>=11`. In practice it is rare that the llvm-project build and a downstream client exchange the hash values.

Reply:
Wouldn't this affect any access to a DenseMap across the boundary for example?

That is any API where LLVM would initialize a DenseMap and return a reference to it to the user which then access elements by computing key hashes?

Reply:
It is definitely a problem for LLVM built using GCC and an external project, based on LLVM, built using CLANG 11.  Is there a reliable check that we can add (e.g. in `llvm/include/llvm/Config/abi-breaking.h.cmake`) to detect this at the build time rather than failing during the execution?

Reply:
LLVM built using GCC are typically release builds and LLVM_ENABLE_ABI_BREAKING_CHECKS is 0. The `else` branch is taken and there is no impact.

If we want to be over-cautious, this condition can be refined to:
`LLVM_ENABLE_ABI_BREAKING_CHECKS && (!defined(__clang__) || __clang_major__ > 11 || defined(__PIC__))`.
We could code the condition to `utils/bazel/llvm_configs/abi-breaking.h.cmake`, but I am hoping that the complexity isn't needed.

Reply:
Maybe it would make sense to move the function to the source file for the LLVM_ENABLE_ABI_BREAKING_CHECKS case? It will add cost, but it's okay for assertion-enabled builds.

Reply:
In my case it is a `Release` build with GCC and `-DLLVM_ENABLE_ASSERTIONS=ON`.

Could you please explain how `__PIC__` check helps?

Reply:
We are still dealing with the fallout from this change in [Triton](https://github.com/triton-lang/triton/pull/4760).

It seems that breaking the ABI across different compilers is acceptable, but it would be good to detect this and error out. We already [detect mismatches](https://github.com/llvm/llvm-project/blob/bc91f3cdd57cbe4b0a456626f52960158cb3232f/llvm/include/llvm/Config/abi-breaking.h.cmake#L45) of `LLVM_ENABLE_ABI_BREAKING_CHECKS`. Would it be acceptable to just add the `clang > 11` condition there?